### PR TITLE
generate: pull WIT files from OCI registry

### PR DESCRIPTION
**Pull request by Mossaka**
integrating `wasm-pkg-tools/wkg` to wit-bindgen-go to pull WIT files from an OCI registry to local. 

Usage: `wit-bindgen-go generate ghcr.io/webassembly/wasi/http:0.2.0`

~~It installs the `wkg` CLI tool to the cache directory and uses it to pull & unpack WIT OCI artifact to a wasm component. Then it uses `wasm-tools` to unpack WIT files from the component tot target directory.~~

~~I've tested this locally with my aarch64 macos. Will do more testing before I make this PR ready.~~

I've tested this on both Apple Silicon MacOS and amd64 Linux. 

~~Update: I removed the installation step so now it calls the `wkg` from PATH to pull WIT files. This requires users to install `wkg` themselves.~~

Update 2: Instead of shelling out `wkg`, I used `regclient` APIs to pull OCI artifact and unpack as a wasm file. This removes the need for an extra CLI to work with OCI. 

## Sep 28 update:

* [cmd/wit-bindgen-go: move loadWitModule to witcli](https://github.com/bytecodealliance/wasm-tools-go/pull/146/commits/862382e93919e2634e31ed1e43e2bf220d28c797) 
* [cmd/wit-bindgen-go/cmd/generate: remove writeWITPackage](https://github.com/bytecodealliance/wasm-tools-go/pull/146/commits/137390e8c0dad75a3427c0908c2f6b8ff3c608a1)

### CODE REVIEW

File: internal/oci/oci.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+// It invokes "regclient" APIs to pull the artifact and then
+// processes it with `wasm-tools`.
+// the WIT files are stored in the `<out>` directory
+func PullWIT(ctx context.Context, path, out string) error {
+	r, err := ref.New(path)
+	if err != nil {
+		return fmt.Errorf("failed to parse ref: %v", err)
+	}
+
+	rc := regclient.New()
```
Comment (thomastaylor312) [inline internal/oci/oci.go:31]: I love the idea of integrating directly and understand wanting to avoid another CLI dependency. However, I would recommend using the same config as `wkg` for fetching and loading credentials (which I don't know if `regclient` currently supports, but that should be a requirement as well for those with private registries): https://github.com/bytecodealliance/wasm-pkg-tools/pull/79 (this will be merged soon, just dropping here so you can see the documentation on the config file and how to fetch from the correct OCI registry using the registry.json file

File: internal/oci/oci.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+	if _, err = io.Copy(outFile, rdr); err != nil {
+		return fmt.Errorf("failed to write to output file: %v", err)
+	}
+
+	wasmTools, err := exec.LookPath("wasm-tools")
+	if err != nil {
+		return err
+	}
+	wasmCmd := exec.Command(wasmTools, "component", "wit", wasmFile, "--out-dir", out)
```
Comment (thomastaylor312) [inline internal/oci/oci.go:80]: So the main wit-bindgen stuff supports bindgen directly from a wit package (what is stored in OCI) rather than a wit file. This is probably fine for now, but the better experience long term would be to parse it out from the wit package instead

**Review by thomastaylor312 — APPROVED** [review]
Just wanted to note how we were hoping for this to integrate in with upstream wasm tooling.

FWIW, one of the next (and I mean soon) features going into `wkg` is going to be moving the `wit` command to it. So you could just install `wkg` and then have it do the equivalent of `wkg fetch && wkg wit`. Completely willing to come to the Go subgroup to discuss the current tooling and roadmap if it helps

**Comment by thomastaylor312** [comment]
(Whoops that was supposed to be a comment and not an approve, but 🤷 )

File: internal/oci/oci.go (hunk 1 above)
Comment (Mossaka) [reply internal/oci/oci.go:31]: Is it possible to integrate `wkg` features directly into wasm-tools? This will benefit this project since it's already shelling out `wasm-tools` for generating components.

Comment (thomastaylor312) [reply internal/oci/oci.go:31]: We'll have to see on that. I'm not opposed, but a lot of people think wasm-tools has gotten far too big/bloated (which I think is a fair assessment). So I'm not completely sure what would happen here

Comment (ydnar) [reply internal/oci/oci.go:31]: It has less functionality than the "go" binary, so not sure I'd call it bloated!

I'd support moving wkg into wasm-tools.

File: internal/oci/oci.go (hunk 2 above)
Comment (ydnar) [reply internal/oci/oci.go:80]: What do you mean directly?

**Comment by Mossaka** [comment]
My understanding for the failure of "Test with WebAssembly" test is that the `oci` package makes a network call to fetch OCI artifacts, which isn't supported by wasip1 Go. Can we make this an exception for the tests, @ydnar ?

File: internal/oci/oci.go (hunk 1 above)
Comment (Mossaka) [reply internal/oci/oci.go:35]: Will the config file be supported by standard OCI tools like `regclient` and `oras`? I am asking this question because I think the responsibility of parsing a standardized config file should belong to tools in that field, and should be out of scope for this repo. If the answer is yes, then I think it's fine for merging this PR and we can land new versions of `regclient` later for it.

File: internal/oci/oci.go (hunk 2 above)
Comment (thomastaylor312) [reply internal/oci/oci.go]: As in it can read the binary wit package rather than needing to convert it to a text representation

File: internal/oci/oci.go (hunk 1 above)
Comment (thomastaylor312) [reply internal/oci/oci.go:35]: The config file I am mentioning is separate from oci or warg and is specific to wasm registry tooling. So it would belong in here and not in oras or regclient. I can clarify what we're doing here and why and what the plan is for standardizing around "locking" the fetched wit packages next week in the SIG Go meeting.

As for the wasm-tools thing, I think we're going to try and have a convo about that in the component model meeting.

File: internal/oci/oci.go (hunk 2 above)
Comment (Mossaka) [reply internal/oci/oci.go]: I think I can remove this line calling to wasm-tools since it can handle a binary WIT file.

File: internal/oci/oci_wasi.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+	"fmt"
+)
+
+// IsOCIPath checks if a given path is an OCI path
+func IsOCIPath(path string) bool {
+	return false
+}
+
+func PullWIT(ctx context.Context, path, out string) error {
+	return fmt.Errorf("OCI not supported on WASI")
```
Comment (ydnar) [inline internal/oci/oci_wasi.go:16]: No need for errorf if this is a constant

File: internal/oci/oci_other.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+		return fmt.Errorf("layer contains invalid digest: %s: %v", string(layer.Digest), err)
+	}
+
+	rdr, err := rc.BlobGet(ctx, r, layer)
+	if err != nil {
+		return fmt.Errorf("failed to fetch blob: %v", err)
+	}
+	defer rdr.Close()
+
+	wasmCmd := exec.Command(wasmTools, "component", "wit", "--out-dir", out)
```
Comment (ydnar) [inline internal/oci/oci_other.go:70]: I don't think it's necessary to emit the file, since this can emit WIT to stdout, which can be parsed directly. Can this also pipe directly to JSON?

Comment (Mossaka) [reply internal/oci/oci_other.go:70]: I thought about this but then I realized TinyGo still wants WIT files to embed and componentize a module. I do agree that piping directly to JSON is an optimization.

Comment (ydnar) [reply internal/oci/oci_other.go:70]: Hm. Since wit-bindgen-go can emit WIT, maybe we have the Go generator write the WIT next to the code?

Comment (Mossaka) [reply internal/oci/oci_other.go:70]: > maybe we have the Go generator write the WIT next to the code?
SGTM. Since it's not part of what this PR want's to achieve, I will perhaps open another one for this?

Comment (Mossaka) [reply internal/oci/oci_other.go]: Simplified the `PullWit` function to return a JSON buffer instead.

File: internal/oci/oci_wasi_tinygo.go [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+package oci
+
+import (
+	"bytes"
+	"context"
+	"errors"
+)
+
+// IsOCIPath checks if a given path is an OCI path
+func IsOCIPath(path string) bool {
```
Comment (ydnar) [inline internal/oci/oci_wasi_tinygo.go:12]: Can this be implemented once? Does package `github.com/regclient/regclient/types/ref` import anything unsupported on TinyGo or WASI?

File: internal/oci/oci_wasi_tinygo.go [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+	return false
+}
+
+func PullWIT(ctx context.Context, path string) (*bytes.Buffer, error) {
+	return nil, errors.New("OCI not supported on WASI or TinyGo")
```
Comment (ydnar) [inline internal/oci/oci_wasi_tinygo.go:17]: What part isn’t supported under TinyGo?

File: testdata/issues/issue163.wit [hunk 7]
```diff
@@ -1,4 +1,4 @@
-package test:variants-unioning-types;
+package issue:issue163;
```
Comment (ydnar) [inline testdata/issues/issue163.wit:1]: Let’s pull this out into a separate PR

File: wit/load.go [hunk 8]
```diff
@@ -55,3 +55,33 @@ func LoadWIT(path string) (*Resolve, error) {
 
 	return DecodeJSON(&stdout)
 }
+
+// LoadWITFromBuffer loads WIT data from a provided buffer by processing it through wasm-tools.
+// It expects the buffer to contain valid WIT data and processes it through `wasm-tools`.
+// The result is returned as a *Resolve.
+func LoadWITFromBuffer(buffer *bytes.Buffer) (*Resolve, error) {
```
Comment (ydnar) [inline wit/load.go:62]: Why does this accept a `bytes.Buffer` instead of `[]byte`?

Asking because I’ve generally seen `bytes.Buffer` used within a function, but not used as a return value or argument.

File: wit/load.go [hunk 9]
```diff
@@ -55,3 +55,33 @@ func LoadWIT(path string) (*Resolve, error) {
...
+	if err != nil {
+		return nil, err
+	}
+
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+
+	// Command to run `wasm-tools component wit -j --all-features`
+	cmd := exec.Command(wasmTools, "component", "wit", "-j", "--all-features")
+	cmd.Stdin = buffer   // Use the buffer content as input
```
Comment (ydnar) [inline wit/load.go:73]: Like this could be a `bytes.Reader` around a `[]byte` argument.

File: cmd/wit-bindgen-go/cmd/generate/generate.go [hunk 10]
```diff
@@ -159,5 +182,32 @@ func action(ctx context.Context, cmd *cli.Command) error {
 		}
 	}
 
+	witDir := filepath.Join(out, "wit")
```
Comment (ydnar) [inline cmd/wit-bindgen-go/cmd/generate/generate.go:185]: This seems like a lot of complexity. What is being written to the `wit` directory? WIT or wasm?

File: cmd/wit-bindgen-go/cmd/generate/generate.go [hunk 11]
```diff
@@ -159,5 +182,32 @@ func action(ctx context.Context, cmd *cli.Command) error {
...
+
+	fmt.Fprintf(os.Stderr, "Generated WIT file: %s\n", witFilePath)
+
+	wasmTools, err := exec.LookPath("wasm-tools")
+	if err != nil {
+		return err
+	}
+	var stderr bytes.Buffer
+
+	wasmCmd := exec.Command(wasmTools, "component", "wit", "--wat", "--all-features", "--output", witFilePath)
```
Comment (ydnar) [inline cmd/wit-bindgen-go/cmd/generate/generate.go:200]: This now fails if `wasm-tools` is not present.

**Review by ydnar — CHANGES_REQUESTED** [review]

File: internal/oci/oci_wasi_tinygo.go (hunk 5 above)
Comment (Mossaka) [reply internal/oci/oci_wasi_tinygo.go:12]: Yes, good catch

File: internal/oci/oci_wasi_tinygo.go (hunk 6 above)
Comment (Mossaka) [reply internal/oci/oci_wasi_tinygo.go:17]:
```bash
tinygo run ./cmd/wit-bindgen-go generate testdata/issues/issue170.wit.json
# github.com/docker/libtrust
../../../go/pkg/mod/github.com/docker/libtrust@v0.0.0-20160708172513-aabc10ec26b7/key_manager.go:100:29: undefined: tls.RequireAndVerifyClientCert
../../../go/pkg/mod/github.com/docker/libtrust@v0.0.0-20160708172513-aabc10ec26b7/key_manager.go:145:30: undefined: tls.RequireAndVerifyClientCert
../../../go/pkg/mod/github.com/docker/libtrust@v0.0.0-20160708172513-aabc10ec26b7/key_manager.go:156:19: undefined: tls.VersionTLS10
../../../go/pkg/mod/github.com/docker/libtrust@v0.0.0-20160708172513-aabc10ec26b7/util.go:121:20: testConn.ConnectionState undefined (type *net.TLSConn has no field or method ConnectionState)
```

File: wit/load.go (hunk 8 above)
Comment (Mossaka) [reply wit/load.go:62]: yeah that's a good point, I will swtich it to `[]byte` instead

File: cmd/wit-bindgen-go/cmd/generate/generate.go (hunk 10 above)
Comment (Mossaka) [reply cmd/wit-bindgen-go/cmd/generate/generate.go:185]: the WIT package is output to `<out>/wit/webassembly.wit.wasm`

File: cmd/wit-bindgen-go/cmd/generate/generate.go (hunk 11 above)
Comment (Mossaka) [reply cmd/wit-bindgen-go/cmd/generate/generate.go:200]: yeah but wouldn't generator fail before this code being executed? the `wasm-tools` will run to convert WIT to `.json`, unless the path is a `.json` file already. Perhaps we only want to generate `.json` WIT file and place it adjacent to the source code?

File: internal/oci/oci_wasi_tinygo.go (hunk 6 above)
Comment (ydnar) [reply internal/oci/oci_wasi_tinygo.go:12]: Ugh right. Ok

File: cmd/wit-bindgen-go/cmd/generate/generate.go (hunk 10 above)
Comment (Mossaka) [reply cmd/wit-bindgen-go/cmd/generate/generate.go:185]: In TinyGo you can just do

`tinygo build -target=wasip2 -wit-package wit/webassembly.wit.wasm -wit-world proxy main.go`

File: cmd/wit-bindgen-go/cmd/generate/generate.go (hunk 11 above)
Comment (ydnar) [reply cmd/wit-bindgen-go/cmd/generate/generate.go]: Wit-bindgen-go can generate WIT from JSON. Would that work?

Comment (Mossaka) [reply cmd/wit-bindgen-go/cmd/generate/generate.go]: Yeah see my latest change - I used Resolve to WIT to generate the wit files.

**Comment by ydnar** [comment]
Maybe time for a rebase?

File: cmd/wit-bindgen-go/cmd/generate/generate.go (hunk 11 above)
Comment (Mossaka) [reply cmd/wit-bindgen-go/cmd/generate/generate.go]: moved this logic to #179

File: internal/oci/oci.go (hunk 1 above)
Comment (Mossaka) [reply internal/oci/oci.go]: I think we will eventually get to implement the wkg config file in Go, but for now, I am okay with using regclient to get the WIT package from an OCI reigistry. Will revisit this later

File: CHANGELOG.md [hunk 12]
```diff
@@ -29,6 +29,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 - Generated structs and structs in package `cm` now include a [`HostLayout` field](https://github.com/golang/go/issues/66408) in order to conform with the [relaxed types proposal](https://github.com/golang/go/issues/66984) for `GOARCH=wasm32`. The `cm.HostLayout` type is an alias for `structs.HostLayout` on Go 1.23 or later, and a polyfill for Go 1.22 or earlier.
 - [#163](https://github.com/bytecodealliance/wasm-tools-go/issues/163): added `cm.F32ToU64()` and `cm.U64ToF32()` for flattening `f32` and `u64` types in the Canonical ABI.
 - Test data from [bytecodealliance/wit-bindgen/tests/codegen](https://github.com/bytecodealliance/wit-bindgen/tree/main/tests/codegen).
+- `wit-bindgen-go generate` now accepts a remote registry reference to pull down a WIT package and generate the Go code. Example: `wit-bindgen-go generate ghcr.io/webassembly/wasi/http:0.2.0`.
```
Comment (ydnar) [inline CHANGELOG.md:32]: Move this to [Unreleased] section?

File: cmd/wit-bindgen-go/cmd/generate/generate.go [hunk 13]
```diff
@@ -82,82 +120,72 @@ func action(ctx context.Context, cmd *cli.Command) error {
...
 	}
 	fmt.Fprintf(os.Stderr, "Package root: %s\n", pkgRoot)
 
-	res, err := witcli.LoadOne(cmd.Bool("force-wit"), cmd.Args().Slice()...)
+	path, err := witcli.LoadPath(cmd.Args().Slice()...)
 	if err != nil {
-		return err
+		return nil, err
 	}
+	world, cm, versioned, forceWIT := cmd.String("world"), cmd.String("cm"), cmd.Bool("versioned"), cmd.Bool("force-wit")
```
Comment (ydnar) [inline cmd/wit-bindgen-go/cmd/generate/generate.go:132]: Can these just be inline below? Do they need to be variables?

File: cmd/wit-bindgen-go/cmd/generate/generate.go [hunk 14]
```diff
@@ -82,82 +120,72 @@ func action(ctx context.Context, cmd *cli.Command) error {
...
-				fmt.Println(string(b))
+			if cfg.dryRun {
+				fmt.Println(string(content))
 				fmt.Println()
 				continue
 			}
 
-			f, err := os.Create(path)
-			if err != nil {
+			if err := os.WriteFile(path, content, cfg.outPerm); err != nil {
```
Comment (ydnar) [inline cmd/wit-bindgen-go/cmd/generate/generate.go:185]: Nice

File: internal/witcli/witcli.go [hunk 15]
```diff
@@ -1,30 +1,49 @@
...
-// An error is returned if len(paths) > 1.
-// If paths is empty, or paths[0] == "" or "-", then it reads from stdin.
+// If path is a OCI path, it pulls from the OCI registry and load WIT
+// from the buffer.
+// If path == "" or "-", then it reads from stdin.
 // If the resolved path doesn’t end in ".json", it will attempt to load
 // WIT indirectly by processing the input through wasm-tools.
 // If forceWIT is true, it will always process input through wasm-tools.
-func LoadOne(forceWIT bool, paths ...string) (*wit.Resolve, error) {
+func LoadOne(ctx context.Context, forceWIT bool, path string) (*wit.Resolve, error) {
```
Comment (ydnar) [inline internal/witcli/witcli.go:20]: Should rename this to `LoadWIT` now!

File: wit/load.go [hunk 16]
```diff
@@ -23,35 +26,67 @@ func LoadJSON(path string) (*Resolve, error) {
...
+	r := getReader(path)
+	return loadWIT(path, r)
+}
+
+// LoadWITFromBuffer loads [WIT] data from a buffer by processing it through [wasm-tools].
+// This will fail if wasm-tools is not in $PATH.
+//
+// [WIT]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md
+// [wasm-tools]: https://crates.io/crates/wasm-tools
+func LoadWITFromBuffer(buffer []byte) (*Resolve, error) {
```
Comment (ydnar) [inline wit/load.go:45]: This can be renamed to `ParseWIT`

File: wit/load.go [hunk 17]
```diff
@@ -23,35 +26,67 @@ func LoadJSON(path string) (*Resolve, error) {
...
-	if err != nil {
+	if err := cmd.Run(); err != nil {
 		fmt.Fprint(os.Stderr, stderr.String())
 		return nil, err
 	}
 
 	return DecodeJSON(&stdout)
 }
+
+func getReader(path string) io.ReadCloser {
```
Comment (ydnar) [inline wit/load.go:87]: This can be just `reader`

File: wit/load.go [hunk 18]
```diff
@@ -23,35 +26,67 @@ func LoadJSON(path string) (*Resolve, error) {
 	return DecodeJSON(f)
 }
 
-// LoadWIT loads [WIT] data from path by processing it through [wasm-tools].
+// LoadWITFromPath loads [WIT] data from path by processing it through [wasm-tools].
```
Comment (ydnar) [inline wit/load.go:29]: Back to `LoadWIT`?

**Comment by Mossaka** [comment]
tests being flaky: https://github.com/bytecodealliance/wasm-tools-go/actions/runs/11172488341/job/31059020059?pr=146#step:9:143

**Review by ydnar — APPROVED** [review]
